### PR TITLE
[multi head attention] implement calcCommonDerivative

### PR DESCRIPTION
**Pull request by lhs8928**
- implement multi head attention calcCommonDerivative
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 1]
```diff
@@ -544,7 +544,190 @@ void MultiHeadAttentionLayer::forwarding(RunLayerContext &context,
...
+   * transpose */
+  d_attention_output.setName("multi_head_attention:attention_output:grad");
+
+  projected_query.reshape(TensorDim(
+    {batch_size * num_heads, 1, query_height, projected_query_dim_prop}));
+  d_projected_query.reshape(TensorDim(
+    {batch_size * num_heads, 1, query_height, projected_query_dim_prop}));
+  projected_key.reshape(
+    TensorDim({batch_size * num_heads, 1, key_height, projected_key_dim_prop}));
+  d_projected_key.reshape(
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:640]: please check if these reshape can be removed.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by songgot — APPROVED** [review]
LGTM

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 1 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:617]: d_projected_key will be used in dot_batched_deriv_wrt_2.
In dot_batched operation the batch size is important so it needs to be reshaped

**Pull request merged**